docs/proposals/config_data.md

```diff
@@ -0,0 +1,262 @@
+
+### Dynamic Configuration
+
+Any long-running system has mutating specification over time. In order to facilitate this functionality,
+``ConfigData`` will be versioned and updates will automatically made available to the container (downward API).
+
+``resourceVersion`` (Found in ``ObjectMetadata``) of the ``ConfigData`` object will be updated by the server every time the object is modified.
+After an update, modifications will be made visible to the consumer container. If the consumer uses the ``Data`` pairs
+only for initialization or during starting process, A rolling-update might be necessary to update the components.
+
+It is then the consumer component's responsilibity to make use of the updated data
+once it is made visible (or perform a rolling-update on consumers of that object).
+
+### Advantages
+
+* Reusable across different components.
+* Easy distribution through the API server.
+* Ability to use ``Kubernetes`` authentication and security policies with configuration objects.
+* By leveraging the power of ``/watch`` API along with the new resource, we can change component configurations dynamically.
+* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
+
+### API Resource
+
+A new resource for ``ConfigData`` will be added to the ``API``:
+
+```go
+	type ConfigData struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		Data map[string]string `json:"data,omitempty"`
+	}
+
+	type ConfigDataList struct {
+		TypeMeta `json:",inline"`
+		ListMeta `json:"metadata,omitempty"`
+
+		Items []ConfigData `json:"items"`
+	}
+```
+``Registry`` interface will be added to accompany the ``ConfigData`` resource.  
+``ConfigData`` information will be stored in ``etcd`` by default.
+
+### Volume Source
+
+A new ``ConfigDataSource`` type of volume source containing the ``ConfigData`` object will be added to the ``VolumeSource`` struct in the API:
+
+```go
+	type ConfigDataVolumeSource struct {
+		ConfigDataName string `json:"configDataName"`
+	}
+```
+
+This volume source will be made available and updated by using the downward API volume plug-in (might require some generalizations).
+
+
+## Creating ConfigData
+``ConfigData`` can be created in two ways explicitly and implicitly.
+
+### Backwards-Compatibility
+In order to smooth the transaction of component configuration from other means to ``ConfigData``
+and preserve backwards-compatibility, user will be able to configure components the old way
+which is going to be transformed into a ``ConfigData`` object implicitly which can be updated and reused.
+
+For example, a ``kubelet`` is configured currently as follows:
+```bash
+kubelet --address=0.0.0.0 --port=8888 --api-servers=10.240.13.14:8998,10.240.51.5:8998 --register-node=true
+```
+
+After this proposal is integrated, when you invoke the same commands,
+a ``ConfigData`` object will be created in the API server.
```

Comment:
The benefit would be that each configured component would have a concrete specification of itself in the API server.

Also, there will be a lot of ConfigData objects created for components, so it might not be a big burden.

If the object was not to be created automatically, what do you have in mind for the configuration workflow?
For example, should the user invoke ``kubectl create <ConfigData.ref>`` after a ConfigData object is created implicitly
while configuring a component.